File: src/wright.py
"""
Wright — the one who does the work.

A wright picks up a task, reads the taste guide, does the work,
and submits it for human evaluation. It works within the container
the human defined.

Not an "agent." A wright. One who works within a craft tradition.
"""

import os
import time
from pathlib import Path
from dataclasses import dataclass
from typing import Optional

from .workspace import Workspace
from .task import TaskStore, Task, TaskStatus
from .taste import TasteStore
from .evaluate import evaluate_file


@dataclass
class WorkResult:
    """What a wright produced."""
    task_id: str
    success: bool
    files_changed: list[str]
    change_ids: list[str]
    evaluation_score: float     # self-eval against principles
    defense: str                # why the choices were made
    message: str                # what the wright did


class LLMError(Exception):
    """An LLM call failed in a known, recoverable way."""
    pass


def _call_llm(prompt: str, model: str = "sonnet") -> str:
    """
    Call an LLM. Thin wrapper — keeps the wright model-agnostic.

    Raises LLMError on rate limits, network failures, or bad responses.
    """
    import anthropic

    try:
        client = anthropic.Anthropic()
        response = client.messages.create(
            model="claude-sonnet-4-6",
            max_tokens=4096,
            messages=[{"role": "user", "content": prompt}],
        )
    except anthropic.RateLimitError as e:
        raise LLMError(f"Rate limited: {e}") from e
    except anthropic.APIConnectionError as e:
        raise LLMError(f"Network error: {e}") from e
    except anthropic.APIStatusError as e:
        raise LLMError(f"API error {e.status_code}: {e.message}") from e

    try:
        return response.content[0].text
    except (IndexError, AttributeError) as e:
        raise LLMError(f"Malformed response — no text content: {e}") from e


def build_prompt(task: Task, file_content: str, taste_guide: str,
                  context: dict[str, str] = None) -> str:
    """Construct the wright's working prompt."""
    context_block = ""
    if context:
        parts = []
        for path, content in context.items():
            parts.append(f"### {path}\n```\n{content}\n```")
        context_block = "## Context Files\n" + "\n\n".join(parts) + "\n"

    return f"""You are a wright — a craftsperson who works within a tradition.

## The Two Questions
Every piece of work must answer:
1. **Why are we making this?** {task.why}
2. **How does it solve it elegantly?** Nothing extra, nothing missing.

## Unix Principles
- Files do one thing
- Functions do one thing
- Readable: code is for humans to understand
- Concise: say what you mean, nothing more
- Max 30 lines per function, max 5 args, max 5 nesting depth

## Taste Guide
{taste_guide}

## Your Task
**Intent:** {task.intent}
**Why:** {task.why}
**Scope:** {task.scope}

{context_block}## Target File Content
```
{file_content or "(new file — create from scratch)"}
```

## Instructions
Make the change described in the intent. Follow the principles. Return the complete file content — no explanations, no markdown fences, just the code."""


def _fail_task(tasks: TaskStore, task_id: str, reason: str) -> WorkResult:
    """Record a visible failure — set status, surface the reason."""
    task = tasks.get(task_id)
    if task:
        task.status = TaskStatus.FAILED
        task.defense = reason
        tasks._update(task)
    return WorkResult(
        task_id=task_id, success=False, files_changed=[],
        change_ids=[], evaluation_score=0.0, defense=reason,
        message=f"Failed: {reason}",
    )


class Wright:
    """
    A wright that can pick up tasks and do work.

    The wright reads the taste guide, understands the intent,
    does the work, self-evaluates, and submits for review.
    """

    def __init__(self, workspace_root: str | Path, wright_id: str = "wright-1"):
        self.root = Path(workspace_root).resolve()
        self.wright_id = wright_id
        self.workspace = Workspace(workspace_root)
        self.tasks = TaskStore(workspace_root)
        self.taste = TasteStore(workspace_root)

    def work(self, task_id: str) -> WorkResult:
        """Pick up a task and do the work."""
        task = self.tasks.claim(task_id, self.wright_id)

        # Read context — strip scope tags like :design
        file_scope = task.scope.split(":")[0] if ":" in task.scope else task.scope
        file_content = self.workspace.read_file(file_scope)
        taste_guide = self.taste.guide()
        context = {}
        for ctx_path in (task.context or []):
            ctx_content = self.workspace.read_file(ctx_path)
            if ctx_content:
                context[ctx_path] = ctx_content

        # Lock the file
        try:
            self.workspace.lock(file_scope, self.wright_id, task.intent)
        except RuntimeError:
            return _fail_task(self.tasks, task_id, f"Could not lock {task.scope}")

        # Call LLM — fail visibly if it breaks
        try:
            prompt = build_prompt(task, file_content, taste_guide, context)
            new_content = _strip_fences(_call_llm(prompt).strip())
            defense_prompt = _build_defense_prompt(task, new_content)
            defense = _call_llm(defense_prompt).strip()
        except LLMError as e:
            self.workspace.unlock(file_scope, self.wright_id)
            return _fail_task(self.tasks, task_id, str(e))

        # Write to staging — never touch the live file
        self.workspace.write_file(
            file_scope, new_content, self.wright_id, task.intent,
            staging=True,
        )

        # Self-evaluate against the staged content
        staged_path = self.workspace._staging_dir / file_scope
        eval_result = evaluate_file(staged_path if staged_path.exists() else self.root / file_scope)

        # Unlock
        self.workspace.unlock(file_scope, self.wright_id)

        # Get change IDs
        changes = self.workspace.recent_changes(limit=1)
        change_ids = [c.id for c in changes]

        # Store defense on task and submit
        task = self.tasks.get(task_id)
        if task:
            task.defense = defense
            self.tasks._update(task)
        self.tasks.submit(task_id, change_ids)

        return WorkResult(
            task_id=task_id,
            success=True,
            files_changed=[file_scope],
            change_ids=change_ids,
            evaluation_score=eval_result.score,
            defense=defense,
            message=f"Completed. Self-eval: {eval_result.score:.2f}",
        )

    def work_next(self) -> Optional[WorkResult]:
        """Pick up the next pending task and work on it."""
        pending = self.tasks.pending()
        if not pending:
            return None
        return self.work(pending[0].id)


def _build_defense_prompt(task: Task, code: str) -> str:
    """Ask the wright to defend its choices."""
    return f"""You just completed a piece of work. Now defend it.

**Task:** {task.intent}
**Why it was needed:** {task.why}

**What you produced:**
```
{code[:3000]}
```

Defend your choices. Not what you did — the diff shows that.
Why this form and not another. Why these specific decisions are right.

2-4 sentences. Conceptual, not technical. Go:"""


def _strip_fences(text: str) -> str:
    """Strip markdown code fences."""
    lines = text.strip().split("\n")
    if lines and lines[0].startswith("```"):
        lines = lines[1:]
    if lines and lines[-1].strip() == "```":
        lines = lines[:-1]
    return "\n".join(lines)

File: crates/ww-wright/src/wright.rs
//! Wright — picks up tasks, does the work, defends its choices.

use std::path::{Path, PathBuf};

use tracing::{info, warn};

use ww_workspace::{Task, TaskStatus, TaskStore, TasteStore, Workspace};

use crate::llm::{LlmClient, LlmError};

#[derive(Debug)]
pub struct WorkResult {
    pub task_id: String,
    pub success: bool,
    pub files_changed: Vec<String>,
    pub defense: String,
    pub message: String,
}

pub struct Wright {
    root: PathBuf,
    id: String,
    workspace: Workspace,
    tasks: TaskStore,
    taste: TasteStore,
    llm: LlmClient,
}

impl Wright {
    pub fn new(root: impl AsRef<Path>, llm: LlmClient) -> Self {
        let root = root.as_ref().to_path_buf();
        let meta_dir = root.join(".workwright");
        Self {
            workspace: Workspace::new(&root),
            tasks: TaskStore::new(&meta_dir),
            taste: TasteStore::new(&meta_dir),
            id: "wright-1".to_string(),
            llm,
            root,
        }
    }

    /// Pick up a task and do the work.
    pub async fn work(&self, task_id: &str) -> WorkResult {
        // Claim
        let task = match self.tasks.claim(task_id, &self.id) {
            Ok(t) => t,
            Err(e) => return self.fail(task_id, &format!("claim failed: {e}")),
        };

        let file_scope = scope_to_path(&task.scope);
        let file_content = self.workspace.read_file(&file_scope).unwrap_or_default();
        let taste_guide = self.taste.guide().unwrap_or_default();

        // Read context files
        let mut context = Vec::new();
        for ctx_path in &task.context {
            if let Some(content) = self.workspace.read_file(ctx_path) {
                context.push((ctx_path.clone(), content));
            }
        }

        // Lock
        if let Err(e) = self.workspace.locks.acquire(&file_scope, &self.id, &task.intent) {
            return self.fail(task_id, &format!("lock failed: {e}"));
        }

        // Call LLM for code
        let prompt = build_prompt(&task, &file_content, &taste_guide, &context);
        let new_content = match self.llm.call(&prompt).await {
            Ok(text) => strip_fences(text.trim()),
            Err(e) => {
                self.workspace.locks.release(&file_scope, &self.id).ok();
                return self.fail(task_id, &format!("llm error: {e}"));
            }
        };

        // Call LLM for defense
        let defense_prompt = build_defense_prompt(&task, &new_content);
        let defense = match self.llm.call(&defense_prompt).await {
            Ok(text) => text.trim().to_string(),
            Err(e) => {
                self.workspace.locks.release(&file_scope, &self.id).ok();
                return self.fail(task_id, &format!("defense llm error: {e}"));
            }
        };

        // Write to staging
        if let Err(e) = self.workspace.write_staged(
            &file_scope,
            &new_content,
            &self.id,
            &task.intent,
        ) {
            self.workspace.locks.release(&file_scope, &self.id).ok();
            return self.fail(task_id, &format!("staging write failed: {e}"));
        }

        // Unlock
        self.workspace.locks.release(&file_scope, &self.id).ok();

        // Get change IDs
        let change_ids: Vec<String> = self
            .workspace
            .changelog
            .recent(1)
            .unwrap_or_default()
            .into_iter()
            .map(|c| c.id)
            .collect();

        // Store defense and submit for review
        if let Ok(Some(mut task)) = self.tasks.get(task_id) {
            task.defense = Some(defense.clone());
            self.tasks.update(&task).ok();
        }
        self.tasks.submit(task_id, change_ids).ok();

        info!(task_id, "wright completed work");

        WorkResult {
            task_id: task_id.to_string(),
            success: true,
            files_changed: vec![file_scope],
            defense,
            message: "Completed.".to_string(),
        }
    }

    /// Pick up the next pending task.
    pub async fn work_next(&self) -> Option<WorkResult> {
        let pending = self.tasks.pending().ok()?;
        let task = pending.first()?;
        Some(self.work(&task.id).await)
    }

    fn fail(&self, task_id: &str, reason: &str) -> WorkResult {
        warn!(task_id, reason, "wright failed");

        // Mark task as failed
        if let Ok(Some(mut task)) = self.tasks.get(task_id) {
            task.status = TaskStatus::Failed;
            task.defense = Some(reason.to_string());
            self.tasks.update(&task).ok();
        }

        WorkResult {
            task_id: task_id.to_string(),
            success: false,
            files_changed: vec![],
            defense: reason.to_string(),
            message: format!("Failed: {reason}"),
        }
    }
}

fn scope_to_path(scope: &str) -> String {
    scope.split(':').next().unwrap_or(scope).to_string()
}

fn build_prompt(
    task: &Task,
    file_content: &str,
    taste_guide: &str,
    context: &[(String, String)],
) -> String {
    let mut ctx_block = String::new();
    for (path, content) in context {
        ctx_block.push_str(&format!("### {path}\n```\n{content}\n```\n\n"));
    }
    let ctx_section = if ctx_block.is_empty() {
        String::new()
    } else {
        format!("## Context Files\n{ctx_block}")
    };

    let file_display = if file_content.is_empty() {
        "(new file — create from scratch)"
    } else {
        file_content
    };

    format!(
        r#"You are a wright — a craftsperson who works within a tradition.

## The Two Questions
Every piece of work must answer:
1. **Why are we making this?** {why}
2. **How does it solve it elegantly?** Nothing extra, nothing missing.

## Unix Principles
- Files do one thing
- Functions do one thing
- Readable: code is for humans to understand
- Concise: say what you mean, nothing more
- Max 30 lines per function, max 5 args, max 5 nesting depth

## Taste Guide
{taste_guide}

## Your Task
**Intent:** {intent}
**Why:** {why}
**Scope:** {scope}

{ctx_section}## Target File Content
```
{file_display}
```

## Instructions
Make the change described in the intent. Follow the principles. Return the complete file content — no explanations, no markdown fences, just the code."#,
        why = task.why,
        intent = task.intent,
        scope = task.scope,
        taste_guide = taste_guide,
        ctx_section = ctx_section,
        file_display = file_display,
    )
}

fn build_defense_prompt(task: &Task, code: &str) -> String {
    let truncated = if code.len() > 3000 { &code[..3000] } else { code };
    format!(
        r#"You just completed a piece of work. Now defend it.

**Task:** {intent}
**Why it was needed:** {why}

**What you produced:**
```
{code}
```

Defend your choices. Not what you did — the diff shows that.
Why this form and not another. Why these specific decisions are right.

2-4 sentences. Conceptual, not technical. Go:"#,
        intent = task.intent,
        why = task.why,
        code = truncated,
    )
}

fn strip_fences(text: &str) -> String {
    let lines: Vec<&str> = text.lines().collect();
    let start = if lines.first().is_some_and(|l| l.starts_with("```")) {
        1
    } else {
        0
    };
    let end = if lines.last().is_some_and(|l| l.trim() == "```") {
        lines.len() - 1
    } else {
        lines.len()
    };
    lines[start..end].join("\n")
}
